src/status.ts

```diff
@@ -1,259 +1,454 @@
-    this._testButton.command = 'cmake.ctest';
-    this._testButton.tooltip = localize('run.ctest.tests.tooltip', 'Run CTest tests');
-    this._debugButton.tooltip = localize('launch.debugger.tooltip', 'Launch the debugger for the selected target');
-    this._debugButton.command = 'cmake.debugTarget';
-    this._launchTargetNameButton.command = 'cmake.selectLaunchTarget';
-    this._launchTargetNameButton.tooltip = localize('select.target.tooltip', 'Select the target to launch');
-    this._reloadBuildButton();
-    this.reloadVisibility();
-  }
-
-  reloadVisibility() {
-    setVisible(this._activeFolderButton, Boolean(this._visible && vscode.workspace.workspaceFolders && vscode.workspace.workspaceFolders.length > 1 && !!this._activeFolderButton.text));
-    const autovis_items = [
-      this._cmakeToolsStatusItem,
-      this._buildButton,
-      this._kitSelectionButton,
-      this._targetButton,
-      this._debugButton,
-      this._launchTargetNameButton,
-    ];
-    for (const item of autovis_items) {
-      setVisible(item, this._visible && !!item.text);
-    }
-    setVisible(this._debugButton,
-               this._visible && !this._hideDebugButton && vscode.extensions.getExtension('ms-vscode.cpptools') !== undefined
-                   && !!this._debugButton.text);
+  get text(): string { return this._text; }
+  set text(v: string) {
+    this._text = v;
+    this.update();
   }
 
-  /**
-   * Whether the status bar items are visible
-   */
-  setVisible(v: boolean) {
-    this._visible = v;
-    this.reloadVisibility();
+  get tooltip(): string|null { return this._tooltip; }
+  set tooltip(v: string|null) {
+    this._tooltip = v;
+    this.update();
   }
-  private _visible: boolean = true;
 
-  private _reloadStatusButton() {
-    this._cmakeToolsStatusItem.text = `CMake: ${this._buildTypeLabel}: ${this._statusMessage}`;
-    this.reloadVisibility();
+  protected set icon(v: string|null) { this._icon = v ? `$(${v})` : null; }
+
+  protected set command(v: string|null) { this.button.command = v || undefined; }
+
+  dispose(): void { this.button.dispose(); }
+  update(): void {
+    const visible = this._isVisible();
+    if (!visible || this._forceHidden) {
+      this.button.hide();
+      return;
+    }
+    const text = this._getText(true);
+    if (text === '') {
+      this.button.hide();
+      return;
+    }
+    this.button.text = text;
+    this.button.tooltip = this._getTooltip() || undefined;
+    this.button.show();
   }
 
-  private _reloadDebugButton() {
-    if (!this._launchTargetNameButton.text) {
-      this._debugButton.text = '$(bug)';
-      this._launchTargetNameButton.hide();
-    } else {
-      this._debugButton.text = `$(bug) ${localize('debug', 'Debug')}`;
-      if (this._visible) {
-        this._launchTargetNameButton.show();
-      }
+  protected getTextNormal(): string { return this._text; }
+  protected getTextShort(): string { return this.getTextNormal(); }
```

Comment:
The CTest button has a '-' in the label in `compact` mode.  That should be removed.

Reply:
changed